### DISCUSSION TITLE
What is the purpose of the secret?

### DISCUSSION THREAD

**Discussion by taybin** [frankie567/starlette-csrf#5] [Q&A]
The secret is never read or used anywhere as far as I can tell.  The random value is generated solely with `secrets.token_urlsafe()`.

**Comment by frankie567** [comment]
It's actually used to instantiate an [itsdangerous serializer](https://itsdangerous.palletsprojects.com/en/2.0.x/) to sign the token value:

https://github.com/frankie567/starlette-csrf/blob/af18979e6b049414de0d68292b5a3897c629df25/starlette_csrf/middleware.py#L31

https://github.com/frankie567/starlette-csrf/blob/af18979e6b049414de0d68292b5a3897c629df25/starlette_csrf/middleware.py#L95

But I was actually wondering if it was really necessary. Since it's an opaque value, I'm not sure signing it adds extra protection.

**Reply by frankie567** [reply]
After checking, it seems that [OWASP does recommend to further encrypt the token](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#double-submit-cookie):

> To enhance the security of this solution include the token in an encrypted cookie - other than the authentication cookie (since they are often shared within subdomains) - and then at the server side match it (after decrypting the encrypted cookie) with the token in hidden form field or parameter/header for AJAX calls. This works because a sub domain has no way to over-write an properly crafted encrypted cookie without the necessary information such as encryption key.

~Honestly, I don't understand why it's better but I'll trust them for such sensitive matters 😅~

Ok, now I get it: by doing this, we are 100% sure that an attacker can't forge a value. If we suppose an attacker could achieve to set both the header and the cookie, and replace it by any value, a simple comparison would defeat the protection.

**Reply by taybin** [reply]
Ugh, I thought I did a search of the file for `secret`, but I probably did `self.secret` and it didn't show up.  I wonder if that means it'd be better to not store the value in `self.secret`.
